synapse/storage/databases/state/store.py

```diff
@@ -288,6 +294,14 @@ def _get_state_for_group_gather_inflight_requests(
                 # to cover our StateFilter and give us the state we need.
                 break
 
+        if state_filter_left_over != StateFilter.none():
+            if len(inflight_requests) >= MAX_INFLIGHT_REQUESTS_PER_GROUP:
+                # There are too many requests for this group.
+                # To prevent even more from building up, we request the whole
+                # state filter to guarantee that we can be reused by any subsequent
+                # requests for this state group.
+                return (), StateFilter.all()
```

Comment:
Probably clearer without nesting. I also tweaked the comments a bit, but I'm unsure if the changes are correct.

```suggestion
        if state_filter_left_over != StateFilter.none() and len(inflight_requests) >= MAX_INFLIGHT_REQUESTS_PER_GROUP:
            # There are too many in-flight requests for this group.
            # To prevent even more from building up, request the whole
            # state to guarantee that it can be reused by any subsequent
            # requests for this state group.
            return (), StateFilter.all()
```